internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// tlsstate manages TLS keypairs on behalf of the daemon's HTTPS server.
+package tlsstate
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/fs"
+	"math/big"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+var (
+	// systemTime can be faked during testing.
+	systemTime = time.Now
+
+	// idCertValidity sets the x509 identity key validity period. The
+	// certificate is managed by the TLS manager as it is not directly
+	// supported by a crypto.Signer interface, as its an implementation
+	// detail related to TLS (not a signing key). Expiry of this
+	// certificate does not change the identity, but changes the client
+	// pinned identity certificate, forcing the client to re-pair with
+	// the server.
+	idCertValidity = 10 * 365 * 24 * time.Hour
+
+	// tlsCertValidity defines how long the in-memory TLS keypairs are
+	// valid before a new keypair is generated. This allows for
+	// TLS session resumption during this period, and keypair re-use
+	// from concurrent requests. The manager restart will always result
+	// in the creation of a new TLS keypair.
+	tlsCertValidity = time.Hour
+
+	// tlsCertRenewWindow defines how long from the actual certificate
+	// expiry we should rather rotate in order to avoid a race between
+	// the expiry and the session handshake completing in time.
+	tlsCertRenewWindow = 60 * time.Second
+
+	// idCertFile is the public x509 certificate, which holds the
+	// identity public key, and self-signed with the identity key.
+	idCertFile = "identity.pem"
+)
+
+type TLSManager struct {
+	// tlsDir is the location of the PEM keypair files.
+	tlsDir string
+	mu     sync.RWMutex
+	// The active in-memory TLS certificate (keypair).
+	tlsCert *tls.Certificate
+	// The identity certificate loaded from disk.
+	idCert *x509.Certificate
+	// The identity key used for signing TLS certificates.
+	signer crypto.Signer
+}
+
+// NewManager create a new TLS keypair manager. The tlsDir must be a
+// directory of which only the leaf element, at most, does not exist. The
+// leaf directory will be created with 0o700 permissions. The signer
+// must implement a crypto.Signer interface. The signer represents the
```

Comment:
This sentence is unnecessary, given the type:

```suggestion
// leaf directory will be created with 0o700 permissions. The signer represents the
```

Reply:
Fixed, thanks you. Resolving.